rx_signal_processing/rx_dsp_chain.cu

```diff
@@ -15,406 +15,238 @@
-#include "utils/protobuf/computationpacket.pb.h"
+#include "utils/protobuf/rxsamplesmetadata.pb.h"
 #include "utils/protobuf/sigprocpacket.pb.h"
 #include "utils/driver_options/driveroptions.hpp"
 #include "utils/signal_processing_options/signalprocessingoptions.hpp"
 #include "utils/shared_memory/shared_memory.hpp"
 
 #include "dsp.hpp"
+#include "filtering.hpp"
+#include "decimate.hpp"
 
-extern "C" {
-    #include "remez.h"
-}
+#ifdef DEBUG
+#define DEBUG_MSG(x) do {std::cerr << x << std::endl;} while (0)
+#else
+#define DEBUG_MSG(x)
+#endif
 
-// REVIEW #30 all filter-building functions could be placed in a separate file?
-// REPLY could be. Could potentially make a filter building class.
+#define ERR_CHK_ZMQ(x) try {x;} catch (zmq::error_t& e) {} //TODO(keith): handle error
 
-std::vector<double> create_normalized_lowpass_filter_bands(float cutoff, float transition_band,
-                        float Fs) { //REVIEW #5 units for params //REPLY this should prolly go in the doc string
-    std::vector<double> filterbands; //REVIEW #3 describe band choices and how this works //REPLY Should this description be here or in our written documentation?
-    filterbands.push_back(0.0);
-    filterbands.push_back(cutoff/Fs);
-    filterbands.push_back((cutoff + transition_band)/Fs);
-    filterbands.push_back(0.5);
 
-    return filterbands;
-}
+int main(int argc, char **argv){
+  GOOGLE_PROTOBUF_VERIFY_VERSION; // Verifies that header and lib are same version.
+
+  //TODO(keith): verify config options.
+  auto driver_options = DriverOptions();
+  auto sig_options = SignalProcessingOptions();
+  auto rx_rate = driver_options.get_rx_rate(); //Hz
+
+  zmq::context_t sig_proc_context(1); // 1 is context num. Only need one per program as per examples
+
+  zmq::socket_t driver_socket(sig_proc_context, ZMQ_PAIR);
+  ERR_CHK_ZMQ(driver_socket.bind("ipc:///tmp/feeds/1"))
+
+
+  //This socket is used to receive metadata about the sequence to process
+  zmq::socket_t radar_control_socket(sig_proc_context, ZMQ_PAIR);
+  ERR_CHK_ZMQ(radar_control_socket.bind("ipc:///tmp/feeds/2"))
+
+  //This socket is used to acknowledge a completed sequence to radar_control
+  zmq::socket_t ack_socket(sig_proc_context, ZMQ_PAIR);
+  ERR_CHK_ZMQ(ack_socket.bind("ipc:///tmp/feeds/3"))
+
+  //This socket is used to send the GPU kernel timing to radar_control to know if the processing
+  //can be done in real-time.
+  zmq::socket_t timing_socket(sig_proc_context, ZMQ_PAIR);
+  ERR_CHK_ZMQ(timing_socket.bind("ipc:///tmp/feeds/4"))
+
+
+  auto gpu_properties = get_gpu_properties();
+  print_gpu_properties(gpu_properties);
+
+  uint32_t first_stage_dm_rate = 0, second_stage_dm_rate = 0, third_stage_dm_rate = 0;
+  //Check for non integer dm rates
+  if (fmod(rx_rate,sig_options.get_first_stage_sample_rate()) > 0.0) {
+    //TODO(keith): handle error
+  } //TODO(keith): not sure these checks will work.
+/*  else if (fmod(sig_options.get_first_stage_sample_rate(),
```

Comment:
I dont think we would be losing any samples? Its just that we can't represent some ratios of samples rates in exact fractional form, so there will be some slight truncation. 

100000.0/(10000.0/3)
30.0
100000.0/(3333.0)
30.003000300030003